### PR TITLE
changed HPACK UT in order to test some edge cases

### PR DESCRIPTION
**Pull request by amitslavin**

### CODE REVIEW

File: http2/hpack/hpack_test.go [hunk 1]
```diff
@@ -72,45 +73,130 @@ func TestDynamicTableSizeEvict(t *testing.T) {
...
+		},
 
-		// C.2.4 Indexed Header Field
-		// https://httpwg.org/specs/rfc7541.html#rfc.section.C.2.4
-		{"C.2.4", []byte("\x82"),
-			[]HeaderField{pair(":method", "GET")},
-			[]HeaderField{}},
+		{"minimum dynamic key", [][]byte{
+			{0x83, 0x86, 0x45, 0x95, 0x62, 0x72, 0xd1, 0x41, 0xfc, 0x1e, 0xca, 0x24, 0x5f, 0x15, 0x85, 0x2a, 0x4b, 0x63, 0x1b, 0x87, 0xeb, 0x19, 0x68, 0xa0, 0xff, 0x41, 0x8b, 0xa0, 0xe4, 0x1d, 0x13, 0x9d, 0x9, 0xb8, 0xd3, 0x6f, 0xb6, 0xcf, 0x5f, 0x8b, 0x1d, 0x75},
+			{0x83, 0x86, 0x45, 0x9b, 0x62, 0xbb, 0xe, 0x93, 0xc7, 0x6c, 0xab, 0xee, 0x4d, 0x85, 0x1d, 0x3b, 0x8b, 0x67, 0x73, 0x10, 0xac, 0x61, 0x2a, 0x49, 0x3e, 0xd2, 0xa2, 0xb3, 0xd4, 0x82, 0xff, 0xc3, 0xc2, 0xc1, 0xc0, 0x7f, 0x0, 0x86, 0x7d, 0xf7, 0xde, 0x71},
```
Comment (guyarb) [inline http2/hpack/hpack_test.go:328]: what did you change here?

File: http2/hpack/hpack_test.go [hunk 2]
```diff
@@ -72,45 +73,130 @@ func TestDynamicTableSizeEvict(t *testing.T) {
...
+		err        []error
 	}{
-		// C.2.1 Literal Header Field with Indexing
-		// https://httpwg.org/specs/rfc7541.html#rfc.section.C.2.1
-		{"C.2.1", dehex("400a 6375 7374 6f6d 2d6b 6579 0d63 7573 746f 6d2d 6865 6164 6572"),
-			[]HeaderField{pair("custom-key", "custom-header")},
-			[]HeaderField{pair("custom-key", "custom-header")},
+		{"happy flow", [][]byte{{0x83, 0x86, 0x45, 0x95, 0x62, 0x72, 0xd1, 0x41, 0xfc, 0x1e, 0xca, 0x24, 0x5f, 0x15, 0x85, 0x2a, 0x4b, 0x63, 0x1b, 0x87, 0xeb, 0x19, 0x68, 0xa0, 0xff, 0x41, 0x8b, 0xa0, 0xe4, 0x1d, 0x13, 0x9d, 0x9, 0xb8, 0xd3, 0x6f, 0xb6, 0xcf, 0x5f, 0x8b, 0x1d, 0x75, 0xd0, 0x62, 0xd, 0x26, 0x3d, 0x4c, 0x4d, 0x65, 0x64, 0x7a, 0x8a, 0x9a, 0xca, 0xc8, 0xb4, 0xc7, 0x60, 0x2b, 0xb4, 0xea, 0xe0, 0x40, 0x2, 0x74, 0x65, 0x86, 0x4d, 0x83, 0x35, 0x5, 0xb1, 0x1f, 0x40, 0x89, 0x9a, 0xca, 0xc8, 0xb2, 0x4d, 0x49, 0x4f, 0x6a, 0x7f, 0x86, 0x7d, 0xf7, 0xdf, 0x8, 0x9e, 0xb7},
+			{0x83, 0x86, 0x45, 0x9b, 0x62, 0xbb, 0xe, 0x93, 0xc7, 0x6c, 0xab, 0xee, 0x4d, 0x85, 0x1d, 0x3b, 0x8b, 0x67, 0x73, 0x10, 0xac, 0x61, 0x2a, 0x49, 0x3e, 0xd2, 0xa2, 0xb3, 0xd4, 0x82, 0xff, 0xc3, 0xc2, 0xc1, 0xc0, 0x7f, 0x0, 0x86, 0x7d, 0xf7, 0xde, 0x71, 0x97, 0xad}},
```
Comment (guyarb) [inline http2/hpack/hpack_test.go:256]: we captured 4 requests of two types (A and B)
Did you check A1->B1->A2->B2, A1->B1->B2, B1->A2->B2, A2->B2 as well?

File: http2/hpack/hpack_test.go [hunk 3]
```diff
@@ -72,45 +73,130 @@ func TestDynamicTableSizeEvict(t *testing.T) {
...
-		{"C.2.2", dehex("040c 2f73 616d 706c 652f 7061 7468"),
-			[]HeaderField{pair(":path", "/sample/path")},
-			[]HeaderField{}},
-
-		// C.2.3 Literal Header Field never Indexed
-		// https://httpwg.org/specs/rfc7541.html#rfc.section.C.2.3
-		{"C.2.3", dehex("1008 7061 7373 776f 7264 0673 6563 7265 74"),
-			[]HeaderField{{"password", "secret", true}},
-			[]HeaderField{}},
+		{"broken second request last byte- - grpc timeout", [][]byte{
```
Comment (guyarb) [inline http2/hpack/hpack_test.go:290]: what happens if you add random bytes before the first request? (the scenario here is that we didn't clean the buffer correctly and have "leftovers" on the buffer we send to the processing unit)

**Review by guyarb — CHANGES_REQUESTED** [review]

File: http2/hpack/hpack_test.go (hunk 1 above)
Comment (amitslavin) [reply http2/hpack/hpack_test.go:328]: left only the part which is related to “:path”, “:authority”, “:method”, “:scheme” which are the interesting parts of the dynamic table, and then showed that still the dynamic table an be created with only those values, even if some part of the request did not captured

File: http2/hpack/hpack_test.go (hunk 3 above)
Comment (amitslavin) [reply http2/hpack/hpack_test.go:290]: there is some options to the point to raised over here, it depends on the values which are left in the buffer, the common thing to all is that for all cases we expect to get errors. There are 3 types of errors which can occur:
1. invalid Huffman-encoded data - which cased due to encoding issue.
2. invalid indexed representation - which cased due to wrong value created in the buffer which index to a non invalid value  
3. truncated headers - which is cased also due to encoding problem that breaks the representation of the header.

Added UT for each type of error which may happen with its needed buffer state

File: http2/hpack/hpack_test.go (hunk 2 above)
Comment (amitslavin) [reply http2/hpack/hpack_test.go:256]: added the test, summing up the results over here:
1. A1->B1->A2->B2 - happy flow, dynamic table is as expected.
2. A1->B1->B2 - dynamic table is created without the missing request (A2) - result- we lose the timestamp.
3. B1->A2->B2 - dynamic table is not created at all due to "invalid indexed representation" - need to deeply understand the issue - might be a problem
4. A2->B2 - same issue as the last one - no dynamic table was created due to the loss of the first request

Although it might not be a problem because it happens only when first request is lost? it might be because of a header which supposed to be created in the table, such as 'content-type' which should be included in all of the requests and then when it is not found in the dynamic table all of the other requests might point to this missing index?
